### PR TITLE
New access control implementation

### PR DESCRIPTION
**Pull request by rexxars**
This is the branch that introduces big changes to the access control implementation in Imbo, as outlined in #319.

Before a new release, we will need to go over the documentation and make sure everything is well documented and up to date with the changes.

We currently have a working implementation with both a MongoDB and an array adapter.

Currently missing:
- Complete documentation
- Updated CLI-tool with support for adding public keys
- Updated clients (we are still backwards-compatible, but current clients can't use a public key that does not match the user)

@christeredvartsen will need to review this before we move further.

### CODE REVIEW

**Comment by rexxars** [comment]
I found an issue not covered by the tests. The shorturl resource is not working correctly at the moment. I'll see if I can add tests for all resources to ensure they are all working as expected.

File: docs/examples/generateAccessToken.php [hunk 1]
```diff
@@ -1,17 +1,23 @@
...
+];
+
+// Add a query parameter for public key if it differs from the user
+if ($user != $publicKey) {
+    $query[] = 'publicKey=' . $publicKey;
+}
 
 // The URI
-$uri = sprintf("http://imbo/users/%s/images/%s?%s", $publicKey, $image, $query);
+$uri = sprintf("http://imbo/users/%s/images/%s?%s", $user, $publicKey, $image, implode('&', $query));
```
Comment (christeredvartsen) [inline docs/examples/generateAccessToken.php]: There are too many arguments here (only three `%s`'s, but 4 arguments).

File: docs/installation/configuration.rst [hunk 2]
```diff
@@ -374,7 +374,7 @@ This adapter simply stores all images on the file system. It has a single parame
 ``dataDir``
     The base path where the images are stored.
 
-This adapter is configured to create subdirectories inside of ``dataDir`` based on the public key of the user and the checksum of the images added to Imbo. The algorithm that generates the path simply takes the three first characters of the public key and creates directories for each of them, then the full public key, then a directory of each of the first characters in the image identifier, and lastly it stores the image in a file with a filename equal to the image identifier itself.
+This adapter is configured to create subdirectories inside of ``dataDir`` based on the user and the checksum of the images added to Imbo. The algorithm that generates the path simply takes the three first characters of the user and creates directories for each of them, then the full user, then a directory of each of the first characters in the image identifier, and lastly it stores the image in a file with a filename equal to the image identifier itself.
```
Comment (christeredvartsen) [inline docs/installation/configuration.rst]: Need to fix this sentence.

"... , then the full user, ..." sounds weird.

File: docs/installation/upgrading.rst [hunk 3]
```diff
@@ -9,6 +9,71 @@ If you did a :ref:`git clone <git-clone>` you could simply do a ``git pull`` to
...
+Below are the changes you need to be aware of when upgrading to Imbo-2.0.0.
+
+.. contents::
+    :local:
+    :depth: 2
+
+Public key is an authentication detail
+++++++++++++++++++++++++++++++++++++++
+
+Versions prior to 2.0.0 had a 1:1 correlation between what a ``user`` and a ``public key``. In 2.0.0, a ``user`` is the entity which images belong to, while a ``public key`` is one part of the authentication scheme. ``Public keys`` each has their own set of permissions, which can grant them access to different resources within different users.
```
Comment (christeredvartsen) [inline docs/installation/upgrading.rst]: In the first sentence there is a `what` to much. As I'm reading the rest of the paragraph I get the understanding that a public key under one user can also have access to content (resources) owned by other users. Is this intentional? The last part of the last sentence is what I'm aiming at when it says "... different users".

File: docs/usage/api.rst [hunk 4]
```diff
@@ -234,18 +234,18 @@ results in:
 .. code-block:: javascript
 
     {
-      "publicKey": "<user>",
+      "id": "<user>",
```
Comment (christeredvartsen) [inline docs/usage/api.rst]: Is there a reason it's called `id` and not `user`, as it's referred to pretty much everywhere else?

File: library/Imbo/Exception.php [hunk 5]
```diff
@@ -25,11 +25,12 @@
     const ERR_UNSPECIFIED = 0;
 
     // Auth errors
-    const AUTH_UNKNOWN_PUBLIC_KEY = 100;
+    const AUTH_UNKNOWN_USER       = 100;
     const AUTH_MISSING_PARAM      = 101;
     const AUTH_INVALID_TIMESTAMP  = 102;
     const AUTH_SIGNATURE_MISMATCH = 103;
     const AUTH_TIMESTAMP_EXPIRED  = 104;
+    const AUTH_UNKNOWN_PUBLIC_KEY = 105;
```
Comment (christeredvartsen) [inline library/Imbo/Exception.php]: Can't see that this is used in Imbo anymore. Any reason why you want to keep it?

File: library/Imbo/Http/Response/Formatter/JSON.php [hunk 6]
```diff
@@ -63,7 +63,7 @@ public function formatStatus(Model\Status $model) {
      */
     public function formatUser(Model\User $model) {
         return $this->encode(array(
-            'publicKey' => $model->getPublicKey(),
+            'id' => $model->getUserId(),
```
Comment (christeredvartsen) [inline library/Imbo/Http/Response/Formatter/JSON.php]: `id` again. Shall we use this in some places, or use `user` all over the place?

File: library/Imbo/Http/Response/Formatter/XML.php [hunk 7]
```diff
@@ -77,7 +77,7 @@ public function formatUser(Model\User $model) {
 <?xml version="1.0" encoding="UTF-8"?>
 <imbo>
   <user>
-    <publicKey>{$model->getPublicKey()}</publicKey>
+    <id>{$model->getUserId()}</id>
```
Comment (christeredvartsen) [inline library/Imbo/Http/Response/Formatter/XML.php]: Same as above with regards to `id` vs `user`.

File: library/Imbo/Http/Response/Formatter/XML.php [hunk 8]
```diff
@@ -227,12 +227,55 @@ public function formatListModel(Model\ListModel $model) {
...
+    /**
+     * {@inheritdoc}
+     */
     public function formatStats(Model\Stats $model) {
         $users = '';
         $numUsers = 0;
 
         foreach ($model->getUsers() as $user => $stats) {
-            $users .= '<user publicKey="' . $user . '">' . $this->formatArray($stats) . '</user>';
+            $users .= '<user id="' . $user . '">' . $this->formatArray($stats) . '</user>';
```
Comment (christeredvartsen) [inline library/Imbo/Http/Response/Formatter/XML.php]: `id` again.

File: library/Imbo/Http/Response/Formatter/XML.php [hunk 9]
```diff
@@ -288,4 +364,38 @@ private function formatArray(array $data) {
...
     }
+
+    /**
+     * Format access rule data array
+     *
+     * @param array $accessRule
+     * @return string
+     */
+    private function formatAccessRuleArray(array $accessRule) {
+        $rule = '<rule id="' . $accessRule['id'] . '">';
```
Comment (christeredvartsen) [inline library/Imbo/Http/Response/Formatter/XML.php]: `id` again. Won't comment it if it appears again. You get the point.

File: library/Imbo/Resource/AccessRules.php [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public static function getSubscribedEvents() {
+        return [
+            'accessrules.get'  => 'getRules',
+            'accessrules.head' => 'getRules',
+            'accessrules.post' => 'updateRules'
```
Comment (christeredvartsen) [inline library/Imbo/Resource/AccessRules.php]: Inconsistent indentation of `=>`. Some other places they aren't lined up like this, like for instance in the class above. Not very critical, but just pointing it out.

File: library/Imbo/Resource/AccessRules.php [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    /**
+     * Update access rules for the specified public key
+     *
+     * @param EventInterface $event The current event
+     */
+    public function updateRules(EventInterface $event) {
+        $accessControl = $event->getAccessControl();
+        if (!($accessControl instanceof MutableAdapterInterface)) {
```
Comment (christeredvartsen) [inline library/Imbo/Resource/AccessRules.php]: Empty line above `if`.

File: library/Imbo/Resource/AccessRules.php [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        $request = $event->getRequest();
+        $publicKey = $request->getRoute()->get('publickey');
+        $data = json_decode($request->getContent(), true);
+
+        if (!is_array($data)) {
+            throw new InvalidArgumentException('No access rule data provided', 400);
+        }
+
+        $accessControl = $event->getAccessControl();
+        foreach ($data as $rule) {
```
Comment (christeredvartsen) [inline library/Imbo/Resource/AccessRules.php]: Empty line above `foreach` (same as `if` in the previous comment). I won't bother repeating this if there is more occurrences of it.

File: docs/installation/upgrading.rst (hunk 3 above)
Comment (rexxars) [inline docs/installation/upgrading.rst]: This is actually intentional, a public key *doesn't* belong to a user. A public key can access resources on many different users, if configured in this way. This enables for instance a global search across images or a public key which allows read-only access to the `images` and `metadata` resource for some (or all) users.

File: docs/usage/api.rst (hunk 4 above)
Comment (rexxars) [inline docs/usage/api.rst]: No. This seems clumsy. I'll change it to `user`.

File: library/Imbo/Http/Response/Formatter/XML.php (hunk 9 above)
Comment (rexxars) [inline library/Imbo/Http/Response/Formatter/XML.php]: Agree with the above points, but this one is about access rules, not users.

**Comment by rexxars** [comment]
The access control implementation is now used as the basis for the 2.0 branch. Closing this.

**Pull request closed without merging**